functionary/ui/views/environments/utils.py

```diff
@@ -0,0 +1,164 @@
+    If the user is not part of the environment, and they are not on the team,
+    returns a None for both return values.
+
+    If the user has an EnvironmentUserRole, and their permissions are
+    equivalent to their TeamUserRole, then the origin of their effective
+    role will always be from the Environment.
+
+    Args:
+        user: User object for the user whose highest role you want to know
+        environment: The target environment
+
+    Returns:
+        role, team|environment: Return a tuple with the role string and the
+            Environment or Team that role came from
+        None, None: Return a tuple filled with None if user was not part of the
+            team and environment
+    """
+
+    if user not in get_env_members(environment) and user not in get_team_members(
+        environment
+    ):
+        return None, None
+
+    if (
+        env_user_role := EnvironmentUserRole.objects.filter(
+            user=user, environment=environment
+        ).first()
+    ) is None:
+        return (
+            TeamUserRole.objects.get(user=user, team=environment.team).role,
+            environment.team,
+        )
+
+    if (
+        team_user_role := TeamUserRole.objects.filter(
+            user=user, team=environment.team
+        ).first()
+    ) is None:
+        return env_user_role.role, environment
+
+    if (
+        ROLE_HEIRARCHY_MAP.get(env_user_role.role).value
+        < ROLE_HEIRARCHY_MAP.get(team_user_role.role).value
+    ):
+        return team_user_role.role, environment.team
+    return env_user_role.role, environment
+
+
+def get_min_role(user: User, environment: Environment) -> str:
+    """Get the minimum role the user has with respect to the environment"""
+    if (
+        env_role := EnvironmentUserRole.objects.filter(
+            user=user, environment=environment
+        ).first()
+    ) is None:
+        return TeamUserRole.objects.get(user=user, team=environment.team).role
+
+    if (
+        team_role := TeamUserRole.objects.filter(
+            user=user, team=environment.team
+        ).first()
+    ) is None:
+        return env_role.role
+
+    return (
+        env_role.role
+        if ROLE_HEIRARCHY_MAP.get(env_role.role)
+        < ROLE_HEIRARCHY_MAP.get(team_role.role)
+        else team_role.role
+    )
+
+
+def get_users(env: Environment) -> list[dict]:
+    """Get list of users who are part of the environment
+
+    Get a list of users who have access to the environment. This includes
+    the members of the team that the environment belongs to. The list will
+    be sorted in decending order based on role.
+
+    Args:
+        env: The environment to get users from
+
+    Returns:
+        users: A list of dictionaries containing all the users who have
+            access to the environment.
+    """
+    env_members = get_env_members(env)
+    team_members = get_team_members(env)
+
+    # Use set to remove duplicate users
+    all_users = set(env_members + team_members)
```

Comment:
I don't know if it changes your desire to have the dedicated `get_env_members` and `get_team_members` functions or not, but you can get at all of these things using queryset filters:

```python
env_members = User.objects.filter(environmentuserroles__environment=env)
team_members = User.objects.filter(teamuserroles__team=env.team)

all_users = env_members | team_members
```

Reply:
I personally would still keep those functions since I think the explicitness of the function names enhances the readability of the code, so other developers don't need to pause and parse what the Django query is doing.